staging/src/k8s.io/endpointslice/topologies/manager.go

```diff
@@ -0,0 +1,240 @@
+}
+
+// NewManager returns an instance of a topology Manager.
+//
+// `topologies` contains the set of topologies that this Mangaer will use and
+// manage. They should have unique names (as returned by their Topology.Names()
+// function). In the case when two topologies have the same name, the first
+// topology will be used.
+func NewManager(logger klog.Logger, defaultTopologyName string, topologies ...Topology) *Manager {
+	topologiesByName := make(map[string]Topology)
+	for _, topology := range topologies {
+		// Check if a topology with the same name already exists. In case of any
+		// conflict, we use the first occurrence.
+		_, ok := topologiesByName[topology.Name()]
+		if ok {
+			logger.Info("Found multilple topologies with the same name, using its first occurrence", "topologyName", topology.Name())
+			continue
+		}
+
+		topologiesByName[topology.Name()] = topology
+	}
+
+	return &Manager{
+		topologiesByName:          topologiesByName,
+		defaultTopologyName:       defaultTopologyName,
+		activeTopologiesByService: make(map[string]map[discoveryv1.AddressType]Topology),
+	}
+}
+
+// PopulateHints populates topology hints on EndpointSlices and returns updated
+// lists of EndpointSlices to create and update.
+//
+// It also updates all topology caches to reflect the latest topology in use.
+// This means that if a service was previously using topology A and is now using
+// topology B, the cache for topology A will be updated to remove any cached
+// hints for that service.
+func (m *Manager) PopulateHints(logger klog.Logger, service *corev1.Service, si *SliceInfo) ([]*discoveryv1.EndpointSlice, []*discoveryv1.EndpointSlice, []*EventBuilder) {
+	activeTopology, activeTopologyExists := m.activeTopologyForService(si.ServiceKey, si.AddressType)
+	desiredTopology, desiredTopologyEnabled, desiredTopologyExists := m.desiredTopologyForService(service)
+
+	// Identify if there is a change in the topology. Change means that the active
+	// topology is not equal to the desired topology.
+	if activeTopologyExists && (!desiredTopologyExists || activeTopology.Name() != desiredTopology.Name()) {
+		activeTopology.ClearCachedHints(logger, si.ServiceKey, si.AddressType)
+	}
+
+	if !desiredTopologyExists {
+		var events []*EventBuilder
+
+		if desiredTopologyEnabled {
+			// This is the case when the topology annotation has a non-disabled value
+			// but the topology does not exist in the topology manager. Send an event
+			// to notify about an unrecognizable topology.
+			desiredTopologyName, _ := TopologyNameFromAnnotations(service.Annotations)
+			logger.Info("Topology not supported, removing hints", "key", si.ServiceKey, "addressType", si.AddressType, "topology", desiredTopologyName)
+			events = append(events, &EventBuilder{
+				EventType: corev1.EventTypeWarning,
+				Reason:    "TopologyAwareHintsDisabled",
+				Message:   FormatWithAddressTypeAndTopologyName(TopologyNotSupported, si.AddressType, desiredTopologyName),
+			})
```

Comment:
This feels like it could become quite noisy, the only alternative I can think of would be to only trigger this event when the Service itself was updated, but that would require plumbing this into a different code path. Don't think it's necessarily a blocker, just would be good to consider different options here.